test/BasicVaultTest.t.sol

```diff
@@ -0,0 +1,147 @@
+
+    function testUserCannotWithdraw(uint amount) public {
+        console2.log("User cannot withdraw with no balance");
+        vm.assume(amount > 0);
+        vm.prank(user);
+        vm.expectRevert("Insufficient balance!");
+        vault.withdraw(amount);
+    }
+   
+    function testUserMintApproveDeposit(uint amount) public {
+        console2.log("User mints tokens and deposits into vault");
+        vm.startPrank(user);
+        vm.assume(amount > 0);
+        wmd.mint(user, amount);
+        wmd.approve(address(vault), amount);
+        vault.deposit(amount);
+        uint balance = vault.balances(user);
+        assertTrue(balance == amount);
+        vm.stopPrank();
+    }
+
+}
+
+abstract contract StateTokensMinted is StateZero {
+    uint userTokens;
+    
+    function setUp() public override virtual {
+        super.setUp();
+        
+        // state transition: user mints 100 tokens
+        userTokens = 100 * 10**18;
+        vm.prank(user);
+        wmd.mint(user, userTokens);
+    }
+
+}
+
+contract StateTokensMintedTest is StateTokensMinted {
+
+    function testUserCannotDepositWithoutApproval() public {
+        console2.log("User cannot deposit tokens into Vault before setting allowance");
+        vm.prank(user);
+        vm.expectRevert("ERC20: Insufficient approval");
+        vault.deposit(userTokens);
+    }
+
+    function testDepositRevertsIfTransferFails() public {
+        console2.log("Deposit transaction should revert if transfer fails");
+        wmd.setFailTransfers(true);
+        vm.startPrank(user);
+        wmd.approve(address(vault), userTokens);
+        vm.expectRevert("Deposit failed!");
+        vault.deposit(userTokens);
+        vm.stopPrank();
+
+    }
+
+    function testUserApproveAndDeposit() public {
+        console2.log("User sets allowance and deposits into Vault");
+        vm.startPrank(user);
+        wmd.approve(address(vault), userTokens);
+        vault.deposit(userTokens);
+        uint balance = vault.balances(user);
+        assertTrue(balance == userTokens);
+        vm.stopPrank();
+    }
+
+}
+
+abstract contract StateTokensDeposited is StateTokensMinted {
+    
+    function setUp() public override virtual {
+        super.setUp();  
+        vm.startPrank(user);
+        wmd.approve(address(vault), userTokens);
+        vault.deposit(userTokens);
+        vm.stopPrank();
+    }
+}
+
+contract StateTokensDepositedTest is StateTokensDeposited {
+    
+    function testWithdrawRevertsIfTransferFails() public {
+        console2.log("Withdraw transaction should revert if transfer fails");
+        wmd.setFailTransfers(true);
+        vm.prank(user);
+        vm.expectRevert("Token transfer frm User to Vendor failed!");
+        vault.withdraw(userTokens);
+    }
+
+    function testUserCannotWithdrawExcessOfDeposit() public {
+        console2.log("User cannot withdraw more than he has deposited");
+        vm.prank(user);
+        vm.expectRevert("Insufficient balance!");
```

Comment:
If you test with two users, each one depositing 100 wmd, and one of them tries to withdraw 200 wmd, you'll notice that you get a Panic instead of the error message that you are expecting. Why?

Reply:
I'm afraid I do not get a Panic - the test still succeeds.
![image](https://user-images.githubusercontent.com/22549197/165652411-a6ee6bcd-f212-4578-9634-dbc369adc29d.png)

I assume you are asking in the context of my initial withdraw function which contains the require statement.

```
    function withdraw(uint amount) external {      
        require(balances[msg.sender] >= amount, "Insufficient balance!");
        balances[msg.sender] -= amount;
        
        bool success = wmdToken.transfer(msg.sender, amount);
        require(success, "Withdrawal failed!");
        emit Withdrawal(msg.sender, amount);
    }
```
_Please note that I corrected the .transfer() function, passing msg.sender, such that the user would receive wmd tokens from the vault._ 

To my abstract setup for this state, user2 is introduced (mint, approve, deposit are done)

```
abstract contract StateTokensDeposited is StateTokensMinted {
    
    address user2;

    function setUp() public override virtual {
        super.setUp();  
        vm.startPrank(user);
        wmd.approve(address(vault), userTokens);
        vault.deposit(userTokens);
        vm.stopPrank();

        user2 = 0x0000000000000000000000000000000000000002;
        vm.label(user2, "user2");
        
        vm.startPrank(user2);
        wmd.mint(user2, userTokens);
        wmd.approve(address(vault), userTokens);
        vault.deposit(userTokens);
        vm.stopPrank();
    }
}
```
Thus at this point, there are 2 users, each have deposited the same amount of wmd tokens (100 each) to the Vault.

Could you please share your thought process?

Reply:
I can't find it now, but it seemed to me at the time that you were checking against the contract balance of end `wmdToken.balanceOf(address(this))` and not the user balance `balances[msg.sender]`. Probably my fault, apologies. It seems correct to me now.